python/ray/serve/_private/router.py

```diff
@@ -83,62 +139,255 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
+                    f"Failed to fetch queue length for replica: {task.exception()}"
+                )
+            else:
+                replica_id, queue_len, accepted = task.result()
+                if accepted and queue_len < lowest_queue_len:
+                    chosen_replica_id = replica_id
+                    lowest_queue_len = queue_len
 
-    def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
-        new_replica_ids = {r.replica_tag for r in running_replicas}
-        if new_replica_ids != self._replica_id_set:
-            self._replica_id_set = new_replica_ids
-            logger.info(
-                "Got updated replicas for deployment "
-                f"{self._deployment_name}: {new_replica_ids}.",
-                extra={"log_to_stderr": False},
+        if chosen_replica_id is None:
+            return None
+
+        # `self._replicas` may have been updated since the candidates were chosen.
+        # In that case, return `None` so a new one is selected.
+        return self._replicas.get(chosen_replica_id, None)
+
+    def fulfill_next_pending_assignment(self, replica: ReplicaWrapper):
+        """Assign the replica to the next pending assignment in FIFO order.
+
+        If a pending assignment has been cancelled, it will be popped from the queue
+        and not assigned.
+        """
+        while len(self._pending_assignment_futures) > 0:
+            fut = self._pending_assignment_futures.popleft()
+            # Pass over futures that have been cancelled.
+            if not fut.done():
+                fut.set_result(replica)
+                break
+
+    async def fulfill_pending_assignments(self):
+        """Repeatedly tries to fulfill a pending assignment with an available replica.
+
+        This is expected to be run inside a task in self._scheduling tasks.
+
+        When a replica is found, this method will exit if the number of scheduling tasks
+        has exceeded the target number. Else it will loop again to schedule another
+        replica.
+        """
+        try:
+            while len(self._scheduling_tasks) <= self.target_num_scheduling_tasks:
+                async for candidates in self.choose_two_replicas_with_backoff():
+                    replica = await self.select_from_candidate_replicas(candidates)
+                    if replica is not None:
+                        self.fulfill_next_pending_assignment(replica)
+                        break
+        except Exception:
+            logger.exception("Unexpected error in fulfill_pending_assignments.")
+        finally:
+            self._scheduling_tasks.remove(asyncio.current_task(loop=self._loop))
+
+    def maybe_start_scheduling_tasks(self):
+        """Start scheduling tasks to fulfill pending assignments if necessary.
+
+        Starts tasks so that there is at least one task per pending assignment
+        (respecting the max number of scheduling tasks).
+        """
+        tasks_to_start = (
```

Comment:
Can you elaborate why we schedule N tasks instead of just 1 for a single replica assignment?